### PR TITLE
nzxt-kraken3: Ensure that only one of the readers requests Z53 status

### PR DESCRIPTION
**Pull request by aleksamagicka**
In kraken3_read(), multiple readers could execute the same commands for Z53 in parallel, if it's time to refresh the sensor data.

This PR introduces a mutex (`z53_status_request_lock`) that's used to let a single reader send the command for requesting a status report, while others (if they exist), only wait until the report is parsed.

### CODE REVIEW

File: nzxt-kraken3.c [hunk 1]
```diff
@@ -214,16 +216,28 @@ static int kraken3_read(struct device *dev, enum hwmon_sensor_types type, u32 at
...
+			if (mutex_trylock(&priv->z53_status_request_lock)) {
+				/* Request status on demand */
+				reinit_completion(&priv->z53_status_processed);
+
+				/* Send command for getting status */
+				ret = kraken3_write_expanded(priv, z53_get_status_cmd,
+							     Z53_GET_STATUS_CMD_LENGTH);
+				if (ret < 0) {
+					mutex_unlock(&priv->z53_status_request_lock);
+					return ret;
```
Comment (aleksamagicka) [inline nzxt-kraken3.c:229]: In the case it gets here, it failed and will return some error code. Shortly after, other readers will exit, but return -ENODATA. What do you think about return -ENODATA here as well?

File: nzxt-kraken3.c [hunk 2]
```diff
@@ -214,16 +216,28 @@ static int kraken3_read(struct device *dev, enum hwmon_sensor_types type, u32 at
...
+				}
+
+				wait_for_completion(&priv->z53_status_processed);
+				mutex_unlock(&priv->z53_status_request_lock);
+			} else {
+				/* Bail if the sensor report is not parsed in due time */
+				if (wait_for_completion_timeout
```
Comment (aleksamagicka) [inline nzxt-kraken3.c:236]: If someone has an idea how to synchronize the `ret < 0` case from above with this, that would be nice. For now I can only come up with a timeout.

**Comment by jonasmalacofilho** [comment]
This is where _not_ working in isolated patches makes sense.

We're clearly still figuring out what needs to be done, and we're in synchronization code, which requires keeing a more global model of the code.

So I propose we work on the entire `kraken3_read`/`kraken3_raw_event` pair in a single PR.

---

As for what's in this PR, I agree that we do need the mutex now, but I also think that `reinit_completion` needs to be moved into the critical section:

> The re-initialization function, [reinit_completion()](https://www.kernel.org/doc/html/latest/driver-api/basics.html#c.reinit_completion), simply resets the ->done field to 0 (“not done”), without touching the waitqueue. **Callers of this function must make sure that there are no racy wait_for_completion() calls going on in parallel.**
> —https://www.kernel.org/doc/html/latest/scheduler/completion.html?highlight=racy

That said, I haven't looked into this PR in much detail yet, as it's hard to keep track of all possible states in two PRs + concurrency + an unsafe C API (where I keep discovering new preconditions, like the one above).

**Comment by aleksamagicka** [comment]
Ah, will move it into the earlier one. Thought that it would be better to separate.

**Comment by jonasmalacofilho** [comment]
Thanks!

(But if you @amezin disagree and actually prefer to always work in isolated patches, even in this sorts of cases, then please do it that way. I just might have a harder time keeping up with the changes).

**Comment by aleksamagicka** [comment]
I'm not tied into either way particularly, I'll follow whatever is easier for you and others to look at.

**Pull request closed without merging**